src/common/web_cache.h

```diff
@@ -0,0 +1,576 @@
+// {Turn-Key} Eviction Algorithm for Web Caches." 21st USENIX
+// Symposium on Networked Systems Design and Implementation (NSDI 24).
+// 2024.
+
+class WebCacheTest;
+
+namespace webcache {
+
+enum class Metric {
+  metrics_start = 84000,
+  fetch_lat,
+  fetch_error,
+  hit,
+  miss,
+  expired,
+  fetch_wait,
+  fetch_get,
+  size,
+  capacity,
+  clear,
+  metrics_stop
+};
+
+template <typename Key, typename Value>
+class WebCache {
+ public:
+  using ValuePtr = std::shared_ptr<const Value>;
+  struct Result {
+    ValuePtr _value;
+    std::error_code _error;
+
+    Result(ValuePtr value, std::error_code error)
+        : _value(value), _error(error) {}
+    explicit Result(ValuePtr value) : _value(value) {}
+    explicit Result(std::error_code error) : _value(nullptr), _error(error) {}
+    Result(const Result&) = delete;
+    Result& operator=(const Result&) = delete;
+    Result(Result&&) = default;
+    Result& operator=(Result&&) = default;
+    ~Result() = default;
+
+    ValuePtr value() const { return _value; };
+    [[nodiscard]] std::error_code error() const { return _error; };
+  };
+
+ protected:
+  struct Node {
+    std::atomic_bool visited;
+    ceph::real_time added_ts;
+    Key const* key;
+    ValuePtr value;
+
+    explicit Node(ValuePtr value)
+        : visited(false),
+          added_ts(ceph::real_clock::now()),
+          key(nullptr),
+          value(value) {};
+    // for testing
+    Node(ValuePtr value, ceph::real_time added_ts)
+        : visited(false), added_ts(added_ts), key(nullptr), value(value) {};
+
+    Node(Node&&) = default;
+    Node& operator=(Node&&) = default;
+    Node(const Node&) = delete;
+    Node& operator=(const Node&) = delete;
+    ~Node() = default;
+
+    friend std::ostream& operator<<(
+        std::ostream& os, const WebCache<Key, Value>::Node& node) {
+      fmt::print(
+          os, "$n[{}->{} v:{} ts:{}]", fmt::ptr(node.key),
+          fmt::ptr(node.value.get()), node.visited.load(), node.added_ts);
+      return os;
+    }
+  };
+
+ private:
+  CephContext* _cct;
+  std::string _name;
+  PerfCounters* _perf;
+  size_t _capacity;
+  std::list<Node*> _sieve_queue;
+  size_t _sieve_hand;
+  std::unordered_map<Key, Node> _lookup;
+  mutable std::shared_mutex _cache_mutex;
+  std::unordered_map<
+      Key, std::shared_ptr<
+               std::pair<std::promise<Result>, std::shared_future<Result>>>>
+      _fetches;
+  mutable std::mutex _fetches_mutex;
+
+ protected:
+  // sieve_evict removes the next node using the SIEVE algorithm from
+  // _sieve_queue and returns a pointer to the evicted node
+  Node* sieve_evict();
+
+  // sieve_expire_erase_unmutexed removes all expired nodes from the
+  // sieve_queue in place. It writes the expired nodes to out_expired
+  static void sieve_expire_erase_unmutexed(
+      std::list<Node*>& sieve_queue, std::chrono::seconds ttl,
+      ceph::real_time now, std::vector<Node*>& out_expired);
```

Comment:
this vector could similarly be an intrusive list to avoid allocations. this list could use the same list hook as `_sieve_queue`, assuming that nodes are removed from `_sieve_queue` before inserting to `out_expired`

Reply:
I love intrusive lists